src/vellum/workflows/context.py

```diff
@@ -0,0 +1,113 @@
+from contextlib import contextmanager
+from functools import wraps
+import inspect
+from optparse import Option
+import threading
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterator,
+    List,
+    Optional,
+    Protocol,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    runtime_checkable,
+)
+
+from vellum.workflows.events.types import ParentContext
+
+ExecutionContext = Dict[str, Any]
+_CONTEXT_KEY = "_execution_context"
+
+local = threading.local()
+
+_context_lock = threading.Lock()
+
+F = TypeVar("F", bound=Callable[..., Any])
+T = TypeVar("T")
+P = TypeVar("P", bound=ParentContext)
+
+
+def get_execution_context() -> ExecutionContext:
+    """Retrieve the current execution context."""
+    return getattr(local, _CONTEXT_KEY, {})
+
+
+def set_execution_context(context: ExecutionContext) -> None:
+    """Set the current execution context."""
+    setattr(local, _CONTEXT_KEY, context)
+
+
+def get_parent_context() -> ParentContext:
+    return cast(ParentContext, get_execution_context().get("parent_context"))
+
+
+@contextmanager
+def execution_context(**kwargs: Any) -> Iterator[None]:
+    """Context manager for handling execution context."""
+    with _context_lock:
+        prev_context = get_execution_context()
+        new_context = {**prev_context, **kwargs}
+        set_execution_context(new_context)
+    try:
+        yield
+    finally:
+        set_execution_context(prev_context)
+
+
+# Defined set of keywords that should be extracted for the wrapper
+EXECUTION_CONTEXT_ONLY_KEYWORDS: List[str] = ["parent_context"]
+EXECUTION_PASSTHROUGH_KEYWORDS: List[str] = []
+
```

Comment:
base wrapper --- 
- takes in current wanted state
- does not handle parent nesting for us